### PR TITLE
Added `set_fullscreen` API to WindowAdapter.

### PR DESCRIPTION
**Pull request by qhua948**
This PR exposes the fullscreen API for both winit and qt to `WindowAdapter` which can be called from client code to set or unset fullscreen status.

Removed inflexible env variable `SLINT_FULLSCREEN` and instead use the new API to set fullscreen status.

Added an example application `examples/fullscreen_toggle` to demo the toggle.

Note that this is a breaking change, if the maintainer believes that the environment variable should be kept, then I am happy to amend this as well.

See #3283

### CODE REVIEW

File: internal/backends/winit/winitwindowadapter.rs
Comment (ogoffart) [inline internal/backends/winit/winitwindowadapter.rs:206]: I'm thinking we should keep the env variable for now in addition, or that'd be a breaking change.

File: internal/backends/qt/qt_window.rs
Comment (ogoffart) [inline internal/backends/qt/qt_window.rs:1709]: This would also clear other states like minimized and maximized.
In fact, the best would be to add or remove the Qt::WindowFullScreen flag from the set of state using windowStates and setWindowStates

**Review by ogoffart — COMMENTED** [review]
Thanks a lot for the PR! It is well done.

I added a few comments inline.
Also I don't think it is necessary to have a full fledged example just for that feature.

File: internal/backends/winit/winitwindowadapter.rs
Comment (qhua948) [reply internal/backends/winit/winitwindowadapter.rs:206]: Done.

File: internal/backends/qt/qt_window.rs
Comment (qhua948) [reply internal/backends/qt/qt_window.rs]: That makes a lot of sense, done.

**Comment by qhua948** [comment]
> The C++ formatting is broken:-/ Could you please consider to fix that?
> 
> Unfortunately github will not push fix suggestions into other people's branches, so unfortunately you can not just accept that :-) And neither can I do it for you as I can not push an update into your branch either.

Just ran clang-formate :3

**Review by ogoffart — APPROVED** [review]
Thanks a lot for the patch!

**Comment by omac777** [comment]
I was looking for examples/fullscreen_toggle/
I didn't find it after I did a git pull.

git describe --tags
v1.3.0-65-ga3e48c495

None of the above is visible to me.
https://github.com/slint-ui/slint/commit/41a3195be1defed9880ddc3b8c8dcee11da85a2e

git clone https://github.com/slint-ui/slint
Cloning into 'slint'...
remote: Enumerating objects: 115264, done.
remote: Counting objects: 100% (2917/2917), done.
remote: Compressing objects: 100% (1332/1332), done.
remote: Total 115264 (delta 1790), reused 2373 (delta 1529), pack-reused 112347
Receiving objects: 100% (115264/115264), 30.64 MiB | 26.29 MiB/s, done.
Resolving deltas: 100% (81610/81610), done.
cissy@cissy-flex5-fc38sb:~$ cd slint/examples/
cissy@cissy-flex5-fc38sb:~/slint/examples$ ls
7guis           ffmpeg             opengl_texture   slide_puzzle
bash            gallery            opengl_underlay  todo
carousel        gstreamer-player   plotter          uefi-demo
CMakeLists.txt  imagefilter        printerdemo      virtual_keyboard
cpp             iot-dashboard      printerdemo_mcu
energy-monitor  mcu-board-support  printerdemo_old
fancy_demo      memory             README.md
cissy@cissy-flex5-fc38sb:~/slint/examples$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean


fullscreen_toggle directory in examples is still not there.  I don't understand.  I thought this was committed and part of master.

**Comment by omac777** [comment]
I just went to the details for this merge/commit.
https://github.com/slint-ui/slint/actions/runs/7456944726/job/20288395478
I don't see anywhere in those details mentioning it building fullscreen_toggle.

Am I blind?  What were those builds and tests using to evaluate that it passed?

**Comment by qhua948** [comment]
The example was removed at the request of the maintainers before the merge, but you should be able to find the commit hash and use that.
----------------------------------------

Mar 6, 2024 23:32:12 David Marceau ***@***.***>:

**Pull request merged**
